src/commands/publish/upload_form/mod.rs

```diff
@@ -0,0 +1,143 @@
+            info!("Rust project detected. Publishing...");
+            let name = krate::Krate::new("./")?.name.replace("-", "_");
+            // TODO: move into build?
+            build_generated_dir()?;
+            concat_js(&name)?;
+
+            let wasm_module = WasmModule {
+                path: format!("./pkg/{}_bg.wasm", name).to_string(),
+                binding: "wasm".to_string(),
+            };
+
+            let script_path = "./worker/generated/script.js".to_string();
+
+            let assets = ProjectAssets {
+                script_path,
+                wasm_modules: vec![wasm_module],
+            };
+
+            build_form(&assets)
+        }
+        ProjectType::JavaScript => {
+            info!("JavaScript project detected. Publishing...");
+            let package = Package::new("./")?;
+
+            let script_path = package.main()?;
+
+            let assets = ProjectAssets {
+                script_path,
+                wasm_modules: Vec::new(),
+            };
+
+            build_form(&assets)
+        }
+        ProjectType::Webpack => {
+            info!("Webpack project detected. Publishing...");
+            // FIXME(sven): shouldn't new
+            let bundle = wranglerjs::Bundle::new();
+
+            let script_path = bundle.script_path();
+
+            let mut wasm_modules = Vec::new();
+
+            if bundle.has_wasm() {
+                let wasm_module = WasmModule {
+                    path: bundle.wasm_path(),
+                    binding: bundle.get_wasm_binding(),
+                };
+                wasm_modules.push(wasm_module)
+            }
+
+            let assets = ProjectAssets {
+                script_path,
+                wasm_modules,
+            };
+
+            build_form(&assets)
+        }
+    }
+}
+
+fn build_form(assets: &ProjectAssets) -> Result<Form, failure::Error> {
+    let mut form = Form::new();
+
+    // The preview service in particular streams the request form, and requires that the
+    // "metadata" part be set first, so this order is important.
+    form = add_metadata(form, assets)?;
+    form = add_files(form, assets)?;
+
+    Ok(form)
+}
+
+fn add_files(mut form: Form, assets: &ProjectAssets) -> Result<Form, failure::Error> {
+    form = form.file(assets.script_name(), assets.script_path())?;
+
+    for wasm_module in &assets.wasm_modules {
+        form = form.file(wasm_module.filename(), wasm_module.path())?;
+    }
+
+    Ok(form)
+}
+
+fn add_metadata(mut form: Form, assets: &ProjectAssets) -> Result<Form, failure::Error> {
+    let metadata_json = serde_json::json!(&Metadata {
+        body_part: assets.script_name(),
+        bindings: assets.bindings(),
+    });
+
+    let metadata = Part::text((metadata_json).to_string())
+        .file_name("metadata.json")
+        .mime_str("application/json")?;
+
+    form = form.part("metadata", metadata);
+
+    Ok(form)
+}
+
+fn filename_from_path(path: &String) -> String {
```

Comment:
this should return a `Result<String, failure::Error>` and the `unwraps` should be replaced with `?`

Reply:
turns out the unwraps were unwrapping `Option` types, rather than `Result` types. do you think this fn should give an error if it gets back a `None`, bc I'm inclined to feel the caller should. maybe give ProjectAssets a `new` method that checks the path for a non-empty filename?

Reply:
I'd probably make this return `Option<String>` in that case, and have the caller adapt the `None` to its own kind of error. I think there's good arguments both ways; I'm interested to hear what @ashleygwilliams says

Reply:
@steveklabnik i'm inclined to do something like this, the reasoning being that `filename_from_path` is kind of just a utility that i feel does not need to be opinionated about what a valid filename is, as it doesn't imply an application of the filename, whereas i could check either before constructing the ProjectAssets struct, or inside the constructor.